rootchain/samuel/samuel.go

```diff
@@ -0,0 +1,580 @@
+	// Start starts the event tracker from the specified block number
+	Start(uint64) error
+
+	// Stop stops the event tracker
+	Stop() error
+
+	// Subscribe creates a rootchain event subscription
+	Subscribe() <-chan rootchain.Event
+}
+
+// samp defines the SAMP interface for SAMUEL
+type samp interface {
+	// AddMessage pushes a Signed Arbitrary Message into the SAMP
+	AddMessage(rootchain.SAM) error
+
+	// Prune prunes out all SAMs based on the specified event index
+	Prune(uint64)
+
+	// Peek returns a ready set of SAM messages, without removal
+	Peek() rootchain.VerifiedSAM
+
+	// Pop returns a ready set of SAM messages, with removal
+	Pop() rootchain.VerifiedSAM
+}
+
+// signer defines the signer interface used for
+// generating signatures
+type signer interface {
+	// Sign signs the specified data,
+	// and returns the signature and the block number at which
+	// the signature was generated
+	Sign([]byte) ([]byte, uint64, error)
+
+	// VerifySignature verifies the signature for the passed in
+	// raw data, and at the specified block number
+	VerifySignature([]byte, []byte, uint64) error
+
+	// Quorum returns the number of quorum validators
+	// for the given block number
+	Quorum(uint64) uint64
+}
+
+// transport defines the transport interface used for
+// publishing and subscribing to gossip events
+type transport interface {
+	// Publish gossips the specified SAM message
+	Publish(*proto.SAM) error
+
+	// Subscribe subscribes for incoming SAM messages
+	Subscribe(func(*proto.SAM)) error
+}
+
+// storage defines the required storage interface for SAMUEL
+// and its modules
+type storage interface {
+	// ReadLastProcessedEvent reads the last processed event data
+	ReadLastProcessedEvent(string) (string, bool)
+
+	// WriteLastProcessedEvent writes the last processed event data
+	WriteLastProcessedEvent(data string, contractAddr string) error
+}
+
+// eventData holds information on event data mapping
+type eventData struct {
+	payloadType  rootchain.PayloadType
+	eventABI     *abi.Event
+	methodABI    *abi.Method
+	localAddress types.Address
+}
+
+// SAMUEL is the module that coordinates activities with the SAMP and Event Tracker
+type SAMUEL struct {
+	eventData eventData
+	logger    hclog.Logger
+
+	eventTracker eventTracker
+	samp         samp
+	storage      storage
+	signer       signer
+	transport    transport
+}
+
+// NewSamuel creates a new SAMUEL instance
+func NewSamuel(
+	configEvent *rootchain.ConfigEvent,
+	logger hclog.Logger,
+	eventTracker eventTracker,
+	samp samp,
+	signer signer,
+	storage storage,
+	transport transport,
+) *SAMUEL {
+	return &SAMUEL{
+		logger:       logger.Named("SAMUEL"),
+		eventData:    initEventData(configEvent),
+		eventTracker: eventTracker,
+		samp:         samp,
+		signer:       signer,
+		storage:      storage,
+		transport:    transport,
+	}
+}
+
+// initEventLookupMap generates the SAMUEL event data lookup map from the
+// passed in rootchain configuration
+func initEventData(
```

Comment:
Comment seems to explain different function

Reply:
I would make this a method of `eventData` (`initFromConfig`)

Reply:
Leftover 🤦‍♂️ 

Fixed:

[77d8a8e](https://github.com/0xPolygon/polygon-edge/pull/717/commits/77d8a8e008a83e517d20627420177ba4368e9dba)